File: chapters/chp1_code/quadric_surface_visualization.c
#include <stdio.h>
#include <math.h>
#include <string.h>

typedef struct {
    double A, B, C, D, E, F, G, H, I, J;
} QuadricSurface;

QuadricSurface create_quadric(double a, double b, double c, double d, double e, double f,
                             double g, double h, double i, double j) {
    QuadricSurface surface = {a, b, c, d, e, f, g, h, i, j};
    return surface;
}

double evaluate_surface(QuadricSurface s, double x, double y) {
    return s.A * x * x + s.B * y * y + s.C * x * y + s.D * x + s.E * y + s.F;
}

const char* classify_surface(QuadricSurface s) {
    double discriminant = 4 * s.A * s.B - s.C * s.C;
    
    if (discriminant > 0) {
        if (s.A > 0 && s.B > 0) {
            return "Elliptic Paraboloid";
        } else if (s.A < 0 && s.B < 0) {
            return "Elliptic Paraboloid (inverted)";
        } else {
            return "Elliptic surface";
        }
    } else if (discriminant < 0) {
        return "Hyperbolic Paraboloid (Saddle)";
    } else {
        return "Parabolic Cylinder or Degenerate";
    }
}

void print_surface_info(QuadricSurface s) {
    printf("Quadric Surface Information:\n");
    printf("Equation: z = %.1fx² + %.1fy² + %.1fxy + %.1fx + %.1fy + %.1f\n",
           s.A, s.B, s.C, s.D, s.E, s.F);
    printf("Surface type: %s\n", classify_surface(s));
    printf("Coefficients: A=%.1f, B=%.1f, C=%.1f, D=%.1f, E=%.1f, F=%.1f\n",
           s.A, s.B, s.C, s.D, s.E, s.F);
}

void generate_mesh(QuadricSurface s, double x_min, double x_max, double y_min, double y_max,
                   int num_points, const char* filename) {
    FILE* file = fopen(filename, "w");
    if (file == NULL) {
        printf("Error opening file %s\n", filename);
        return;
    }
    
    fprintf(file, "x,y,z\n");
    
    double dx = (x_max - x_min) / (num_points - 1);
    double dy = (y_max - y_min) / (num_points - 1);
    
    for (int i = 0; i < num_points; i++) {
        for (int j = 0; j < num_points; j++) {
            double x = x_min + i * dx;
            double y = y_min + j * dy;
            double z = evaluate_surface(s, x, y);
            fprintf(file, "%.6f,%.6f,%.6f\n", x, y, z);
        }
    }
    
    fclose(file);
    printf("Surface data written to %s\n", filename);
}

void analyze_critical_points(QuadricSurface s, double x_min, double x_max, 
                           double y_min, double y_max) {
    // For quadric surfaces, critical points occur where partial derivatives are zero
    // ∂z/∂x = 2Ax + Cy + D = 0
    // ∂z/∂y = 2By + Cx + E = 0
    
    double det = 4 * s.A * s.B - s.C * s.C;
    
    if (fabs(det) > 1e-10) {
        // Unique critical point
        double x_crit = (s.C * s.E - 2 * s.B * s.D) / det;
        double y_crit = (s.C * s.D - 2 * s.A * s.E) / det;
        
        if (x_crit >= x_min && x_crit <= x_max && 
            y_crit >= y_min && y_crit <= y_max) {
            double z_crit = evaluate_surface(s, x_crit, y_crit);
            printf("Critical point found: (%.3f, %.3f, %.3f)\n", x_crit, y_crit, z_crit);
            
            // Classify critical point using second derivative test
            double fxx = 2 * s.A;
            double fyy = 2 * s.B;
            double fxy = s.C;
            
            double discriminant = fxx * fyy - fxy * fxy;
            
            if (discriminant > 0) {
                if (fxx > 0) {
                    printf("Type: Local minimum\n");
                } else {
                    printf("Type: Local maximum\n");
                }
            } else if (discriminant < 0) {
                printf("Type: Saddle point\n");
            } else {
                printf("Type: Inconclusive (second derivative test fails)\n");
            }
        } else {
            printf("Critical point exists but outside specified range\n");
        }
    } else {
        printf("No unique critical point (degenerate case)\n");
    }
}

void analyze_level_curves(QuadricSurface s) {
    printf("\nLevel Curves Analysis:\n");
    double levels[] = {-2, -1, 0, 1, 2};
    int num_levels = sizeof(levels) / sizeof(levels[0]);
    
    for (int i = 0; i < num_levels; i++) {
        double level = levels[i];
        printf("\nLevel curve z = %.1f:\n", level);
        printf("Equation: %.1fx² + %.1fy² = %.1f\n", s.A, s.B, level - s.F);
        
        double rhs = level - s.F;
        if (rhs > 0 && s.A > 0 && s.B > 0) {
            double a = sqrt(rhs / s.A);
            double b = sqrt(rhs / s.B);
            printf("Ellipse with semi-axes a=%.3f, b=%.3f\n", a, b);
        } else if (rhs == 0) {
            printf("Point (origin)\n");
        } else {
            printf("No real solutions (empty set)\n");
        }
    }
}

int main() {
    printf("Quadric Surface Analysis (C version)\n");
    printf("====================================\n\n");
    
    // Example 1: Elliptic Paraboloid
    printf("Example 1: Elliptic Paraboloid\n");
    QuadricSurface surface1 = create_quadric(1, 1, 0, 0, 0, -1, 0, 0, 0, 0);
    print_surface_info(surface1);
    analyze_critical_points(surface1, -3, 3, -3, 3);
    generate_mesh(surface1, -2, 2, -2, 2, 50, "elliptic_paraboloid_c.csv");
    
    printf("\nSample points:\n");
    for (double x = -1; x <= 1; x += 0.5) {
        for (double y = -1; y <= 1; y += 0.5) {
            printf("f(%.1f, %.1f) = %.3f\n", x, y, evaluate_surface(surface1, x, y));
        }
    }
    
    // Example 2: Hyperbolic Paraboloid (Saddle)
    printf("\n================================================\n");
    printf("Example 2: Hyperbolic Paraboloid (Saddle)\n");
    QuadricSurface surface2 = create_quadric(1, -1, 0, 0, 0, 0, 0, 0, 0, 0);
    print_surface_info(surface2);
    analyze_critical_points(surface2, -3, 3, -3, 3);
    generate_mesh(surface2, -2, 2, -2, 2, 50, "hyperbolic_paraboloid_c.csv");
    
    // Example 3: Complex surface
    printf("\n================================================\n");
    printf("Example 3: Complex Surface with Cross Term\n");
    QuadricSurface surface3 = create_quadric(2, 3, 1, -1, 2, -5, 0, 0, 0, 0);
    print_surface_info(surface3);
    analyze_critical_points(surface3, -3, 3, -3, 3);
    generate_mesh(surface3, -3, 3, -3, 3, 75, "complex_surface_c.csv");
    
    analyze_level_curves(surface1);
    
    printf("\nNote: Generated CSV files can be visualized using:\n");
    printf("- Python with matplotlib\n");
    printf("- R with plot3d\n");
    printf("- Excel or Google Sheets with 3D plotting\n");
    printf("- Any scientific plotting software\n");
    
    return 0;
}

File: chapters/chp1_code/quadric_surface_visualization.rb
#!/usr/bin/env ruby

class QuadricSurface
  attr_reader :A, :B, :C, :D, :E, :F, :G, :H, :I, :J
  
  def initialize(a = 1, b = 1, c = 0, d = 0, e = 0, f = -1, g = 0, h = 0, i = 0, j = 0)
    @A, @B, @C, @D, @E, @F, @G, @H, @I, @J = a, b, c, d, e, f, g, h, i, j
  end
  
  def evaluate(x, y)
    @A * x**2 + @B * y**2 + @C * x * y + @D * x + @E * y + @F
  end
  
  def surface_type
    discriminant = 4 * @A * @B - @C**2
    
    if discriminant > 0
      if @A > 0 && @B > 0
        "Elliptic Paraboloid"
      elsif @A < 0 && @B < 0
        "Elliptic Paraboloid (inverted)"
      else
        "Elliptic surface"
      end
    elsif discriminant < 0
      "Hyperbolic Paraboloid (Saddle)"
    else
      "Parabolic Cylinder or Degenerate"
    end
  end
  
  def equation_string
    terms = []
    terms << "#{@A}x²" if @A != 0
    terms << "#{@B}y²" if @B != 0
    terms << "#{@C}xy" if @C != 0
    terms << "#{@D}x" if @D != 0
    terms << "#{@E}y" if @E != 0
    terms << @F.to_s if @F != 0
    
    equation = terms.empty? ? "0" : terms.join(" + ").gsub("+ -", "- ")
    "z = #{equation}"
  end
  
  def generate_mesh(x_min, x_max, y_min, y_max, num_points, filename)
    File.open(filename, 'w') do |file|
      file.puts "x,y,z"
      
      dx = (x_max - x_min) / (num_points - 1).to_f
      dy = (y_max - y_min) / (num_points - 1).to_f
      
      num_points.times do |i|
        num_points.times do |j|
          x = x_min + i * dx
          y = y_min + j * dy
          z = evaluate(x, y)
          file.puts "#{x},#{y},#{z}"
        end
      end
    end
    
    puts "Surface data written to #{filename}"
  end
  
  def analyze_critical_points(x_range, y_range)
    # For quadric surfaces, critical points occur where partial derivatives are zero
    # ∂z/∂x = 2Ax + Cy + D = 0
    # ∂z/∂y = 2By + Cx + E = 0
    
    det = 4 * @A * @B - @C**2
    
    if det != 0
      # Unique critical point
      x_crit = (@C * @E - 2 * @B * @D) / det
      y_crit = (@C * @D - 2 * @A * @E) / det
      
      if x_crit >= x_range[0] && x_crit <= x_range[1] && 
         y_crit >= y_range[0] && y_crit <= y_range[1]
        z_crit = evaluate(x_crit, y_crit)
        puts "Critical point found: (#{x_crit.round(3)}, #{y_crit.round(3)}, #{z_crit.round(3)})"
        
        # Classify critical point using second derivative test
        fxx = 2 * @A
        fyy = 2 * @B
        fxy = @C
        
        discriminant = fxx * fyy - fxy**2
        
        if discriminant > 0
          if fxx > 0
            puts "Type: Local minimum"
          else
            puts "Type: Local maximum"
          end
        elsif discriminant < 0
          puts "Type: Saddle point"
        else
          puts "Type: Inconclusive (second derivative test fails)"
        end
      else
        puts "Critical point exists but outside specified range"
      end
    else
      puts "No unique critical point (degenerate case)"
    end
  end
  
  def print_info
    puts "Quadric Surface Information:"
    puts "Equation: #{equation_string}"
    puts "Surface type: #{surface_type}"
    puts "Coefficients: A=#{@A}, B=#{@B}, C=#{@C}, D=#{@D}, E=#{@E}, F=#{@F}"
  end
end

puts "Quadric Surface Analysis (Ruby version)"
puts "======================================="
puts

# Example 1: Elliptic Paraboloid
puts "Example 1: Elliptic Paraboloid"
surface1 = QuadricSurface.new(1, 1, 0, 0, 0, -1)
surface1.print_info
surface1.analyze_critical_points([-3, 3], [-3, 3])
surface1.generate_mesh(-2, 2, -2, 2, 50, "elliptic_paraboloid_ruby.csv")

puts "\nSample points:"
(-1..1).step(0.5) do |x|
  (-1..1).step(0.5) do |y|
    puts "f(#{x}, #{y}) = #{surface1.evaluate(x, y)}"
  end
end

# Example 2: Hyperbolic Paraboloid (Saddle)
puts "\n" + "="*50
puts "Example 2: Hyperbolic Paraboloid (Saddle)"
surface2 = QuadricSurface.new(1, -1, 0, 0, 0, 0)
surface2.print_info
surface2.analyze_critical_points([-3, 3], [-3, 3])
surface2.generate_mesh(-2, 2, -2, 2, 50, "hyperbolic_paraboloid_ruby.csv")

# Example 3: More complex surface
puts "\n" + "="*50
puts "Example 3: Complex Surface with Cross Term"
surface3 = QuadricSurface.new(2, 3, 1, -1, 2, -5)
surface3.print_info
surface3.analyze_critical_points([-3, 3], [-3, 3])
surface3.generate_mesh(-3, 3, -3, 3, 75, "complex_surface_ruby.csv")

# Level curves analysis
puts "\n" + "="*50
puts "Level Curves Analysis for Surface 1:"
levels = [-2, -1, 0, 1, 2]
levels.each do |level|
  puts "\nLevel curve z = #{level}:"
  puts "Equation: #{surface1.A}x² + #{surface1.B}y² = #{level - surface1.F}"
  
  if level - surface1.F > 0
    a = Math.sqrt((level - surface1.F) / surface1.A)
    b = Math.sqrt((level - surface1.F) / surface1.B)
    puts "Ellipse with semi-axes a=#{a.round(3)}, b=#{b.round(3)}"
  elsif level - surface1.F == 0
    puts "Point (origin)"
  else
    puts "No real solutions (empty set)"
  end
end

puts "\nNote: Generated CSV files can be visualized using:"
puts "- Python with matplotlib"
puts "- R with plot3d"
puts "- Excel or Google Sheets with 3D plotting"
puts "- Any scientific plotting software"